### PR TITLE
RavenDB-3994 we should execute index update triggers when removing from MapReduceIndex using RemoveFromIndexTask

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

**Comment by ayende** [comment]
@ppekrol We need CLA from Lars for the changes to scripted index results.

**Comment by ppekrol** [comment]
1. Different flow, part of multi-level reduction process
2. Yes, it works with side-by-side. Side-by-side indexing is basically same indexing process with additional thresholds that at certain point (time/etag/etc) will swap both indexes.

File: Raven.Database/Bundles/ScriptedIndexResults/ScriptedIndexResultsIndexTrigger.cs [hunk 1]
```diff
@@ -153,7 +166,7 @@ public override void Dispose()
                                     }
                                     catch (Exception e)
                                     {
-                                        Log.Warn("Could not apply index script " + scriptedIndexResults.Id + " to index result with key: " + kvp.Key, e);
+                                        Log.WarnException("Could not apply index script " + scriptedIndexResults.Id + " to index result with key: " + kvp.Key, e);
```
Comment (lcorneliussen) [inline Raven.Database/Bundles/ScriptedIndexResults/ScriptedIndexResultsIndexTrigger.cs]: Funny. Was just about creating a PullRequest for this.

**Comment by lcorneliussen** [comment]
1) I guess the side-by-side index does not run the scripts configured for the original index, right?
2) When swapping out the original with the newly built index, what happens to scriptedindexresults? Will it run delete on everything in the old index, and then index for everything in the new? Or will it compare?

**Comment by ayende** [comment]
1) No, it doesn't.
2) It will keep the existing stuff, unless you have manually reset it.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by lcorneliussen** [comment]
A manual reset runs delete, right?

Could be a nice addition to Scripted Index Results to have persistent projections - that is basically what  I do with the scripts. Would then be great, if it would work with side-by-side and then only replace actual changes...

Now we have no way of rebuilding certain stages in the multi-map-reduce in a performant way.

**Comment by ayende** [comment]
Yes, manual reset doesn't run delete.
The problem is that your script can do anything, like modify documents.
And we can't track what you did and then fixup

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811



On Fri, Nov 6, 2015 at 10:08 AM, Lars Corneliussen <notifications@github.com

**Comment by lcorneliussen** [comment]
When will this be part of next stable?

(I still do not get under which conditions this happens... But it happens again... :unamused:

**Comment by HibernatingRhinosServiceAccount** [comment]
We will release a stable today

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811



On Thu, Nov 19, 2015 at 3:23 PM, Lars Corneliussen <notifications@github.com

**Comment by lcorneliussen** [comment]
Please can you explain when this bug occurs, so that I can work around it somehow? I won't be able to upgrade to the new stable right away...

Or at least understand it? It happend once in 14 days today - so it must be some kind of special case, since it is a deterministic code path....

**Comment by ppekrol** [comment]
@lcorneliussen 
This happens during deletion of the document, but only if the document key matches the reduction key. 

Please look at the test:
https://github.com/ayende/ravendb/blob/master/Raven.Tests.Issues/RavenDB_3994.cs#L87

We are grouping by the Id, so it implies that the reduction key will be the same as the document key and when such a delete is encountered then we were not firing the ScriptedIndexResults scripts.

**Comment by lcorneliussen** [comment]
Wrote some comments on the code/commit directly.
I think I wrote the test you point at, so that doesn't help :-)

I'm interested in what actually was fixed in the code. If that is hard to explain, ok, but if you can't that doesn't really make me more confident.

**Comment by ppekrol** [comment]
@lcorneliussen 
For the indexing/reducing process to work properly (stay consistent), when we are removing an index entry we must execute all AbstractIndexUpdateTriggerBatcher implementations. In your case the implementation comes from the ScriptedIndexResults and can be found here: https://github.com/ayende/ravendb/blob/master/Raven.Database/Bundles/ScriptedIndexResults/ScriptedIndexResultsIndexTrigger.cs#L47

The fix was based on the test that you supplied and covered a scenario with removing index entries that had reduction key equal to your document key.

**Comment by lcorneliussen** [comment]
I compared the stack traces of a remove followed by key != reducekey and key == reducekey:
https://www.diffchecker.com/keoas63s

Looks like xpluto/pluto runs ExecuteIndexing and pluto/pluto runs directly into voron-code...

Will check more tomorrow.

**Comment by lcorneliussen** [comment]
Did a root cause analysis. IMHO, the fix here is wrong.
New Issue: http://issues.hibernatingrhinos.com/issue/RavenDB-4045

**Pull request merged**